org/mozilla/jss/tests/BenchmarkSSLSocket.java

```diff
@@ -0,0 +1,245 @@
+package org.mozilla.jss.tests;
+
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.security.KeyStore;
+import java.security.SecureRandom;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLServerSocketFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509ExtendedKeyManager;
+
+import org.mozilla.jss.CryptoManager;
+import org.mozilla.jss.provider.javax.crypto.JSSKeyManager;
+import org.mozilla.jss.provider.javax.crypto.JSSNativeTrustManager;
+
+/**
+ * Utility for benchmarking the performance of SSLSocket implementations.
+ *
+ * This benchmark is a server-side SSLSocket benchmark for use with clients
+ * which measure performance. This benchmark does not, itself, make any
+ * performance measurements. For instance:
+ * <pre>
+ * $ ./run_test.sh org.mozilla.jss.tests.BenchmarkSSLSocket JSS.legacy Server_RSA 8443 1024 &
+ * $ siege -c 100 -b -t 5m https://localhost:8443
+ * $ kill %1
+ * </pre>
+ *
+ * There are three supported SSLSocket implementations:
+ *
+ *  1. org.mozilla.jss.ssl.SSLSocket, JSS's legacy implementation
+ *     name: JSS.legacy
+ *  2. org.mozilla.jss.ssl.javax.JSSSocket, JSS's new javax implementation
+ *     name: JSS.SSLSocket
+ *  3. sun.security's SSLSocketImpl from the current JDK.
+ *     name: SunJSSE.SSLSocket
+ *
+ * It is suggested to disable all logging (for instance, via:
+ * truncate -s 0 tools/logging.properties) in order have reproducible
+ * results.
+ *
+ * This class takes four arguments when invoked:
+ *
+ *  1. The name of the implementation to benchmark, see above.
+ *  2. An alias of the certificate or path to a PKCS#12 file. Only SunJSSE
+ *     accepts a PKCS#12 as path -- the two JSS based SSLSocket
+ *     implementations will utilize a nickname instead.
+ *  3. The port to listen on.
+ *  4. The size of the HTTP message to fake.
+ *
+ * Note that, when utilizing a JSS provider, JSS must be loaded via a
+ * java.security. When utilizing SunJSSE, for best results, do not load
+ * JSS via java.security.
+ *
+ * It is suggested to use run_test.sh from the build/ directory for this.
```

Comment:
It might be better to put the above doc in a Markdown page for better visibility.

Reply:
Hmmm... We could. But then the documentation is separate from the code and we'd have to remember to update each. This also helps people who stumble upon the code understand how to use it without having to find the relevant doc. :-) Preferences?

Reply:
I think if the doc is about how to use the class as a programmer, it can stay here as Javadoc. But if it's about using it from command line, it belongs to a separate user's guide. We can put a link from here to the doc, and probably from the doc to here as well. Would that be sufficient? Man pages are separate from the code too.

Reply:
We probably can include the test results that you have above in the doc as well so we have some idea about the current performance.